framework/graph/GraphBuilder.kt

```diff
@@ -127,33 +128,84 @@ class GraphBuilder(
         lock.readLock().withLock { edges += edge }
     }
 
-    fun addEdgeCandidate(edge: EdgeCandidate) {
-        val key = when (edge) {
-            is EdgeCandidate.Has -> edge.targetIID
-            is EdgeCandidate.Isa -> edge.targetLabel
-            is EdgeCandidate.Owns -> edge.targetLabel
-            is EdgeCandidate.Plays -> edge.sourceLabel
-            is EdgeCandidate.Sub -> edge.targetLabel
+    fun addThingEdgeCandidate(edgeCandidate: ThingEdgeCandidate) {
+        val key = when (edgeCandidate) {
+            is ThingEdgeCandidate.Has -> edgeCandidate.targetIID
+            is ThingEdgeCandidate.Isa -> edgeCandidate.targetLabel
         }
-        edgeCandidates.compute(key) { _, existing -> if (existing == null) listOf(edge) else existing + edge }
+        thingEdgeCandidates.compute(key) { _, existing -> if (existing == null) listOf(edgeCandidate) else existing + edgeCandidate }
     }
 
     private fun completeEdges(missingVertex: Vertex) {
-        val key = when (missingVertex) {
-            is Vertex.Type -> missingVertex.type.label.name()
-            is Vertex.Thing -> missingVertex.thing.iid
-        }
-        edgeCandidates.remove(key)?.let { candidates ->
-            candidates.forEach { edges += it.toEdge(missingVertex) }
+        if (missingVertex is Vertex.Thing) {
+            val key = missingVertex.thing.iid
+            thingEdgeCandidates.remove(key)?.let { candidates ->
+                candidates.forEach { edges += it.toEdge(missingVertex) }
+            }
         }
     }
 
     fun completeAllEdges() {
         // Since there is no protection against an edge candidate, and the vertex that completes it, being added
         // concurrently, we do a final sanity check once all vertices + edges have been loaded.
-        lock.readLock().withLock { (graph.thingVertices + graph.typeVertices).values.forEach { completeEdges(it) } }
+        lock.readLock().withLock {
+            graph.thingVertices.values.forEach { completeEdges(it) }
+            renderSchema()
+        }
     }
 
+    private fun renderSchema() {
+        val vertices = allTypeVertices // all schema vertices
+        val visited: MutableMap<Vertex, Set<Pair<String, Vertex.Type>>> = mutableMapOf()
+
+        vertices.values.forEach {
+            renderEdges(it.type, visited)
+        }
+    }
+
+    private fun renderEdges(type: Type, visited: MutableMap<Vertex, Set<Pair<String, Vertex.Type>>>): Set<Pair<String, Vertex.Type>> {
+        if (type.isRoot) return emptySet()
+
+        val remoteType = type.asRemote(transactionState.transaction)
+        val superType = remoteType.supertype!!
+        if (!allTypeVertices.containsKey(type.label.name())) return renderEdges(superType, visited)
+
+        val vertex = allTypeVertices[type.label.name()]!!
```

Comment:
`typeVertex` just to make it clear this is related to the `type` argument

Reply:
Done.